api/processors/group.py

```diff
@@ -0,0 +1,86 @@
+from datetime import datetime
+
+import pytz
+from django.db.models import F
+
+from api.models import CoordinatorLogEvent, Group, GroupLogEvent, StudentLogEvent, TeacherLogEvent
+from api.models.choices.log_event_types import (
+    CoordinatorLogEventType,
+    GroupLogEventType,
+    StudentLogEventType,
+    TeacherLogEventType,
+)
+from api.models.choices.statuses import (
+    CoordinatorStatus,
+    GroupStatus,
+    StudentStatus,
+    TeacherStatus,
+)
+from api.models.constants import CoordinatorGroupLimit
+from api.processors.base import Processor
+
+
+class GroupProcessor(Processor):
+    @classmethod
+    def start(cls, group: Group) -> None:
+        cls._create_log_events_start(group)
+        cls._set_statuses_start(group)
+
+    @classmethod
+    def _create_log_events_start(cls, group: Group) -> None:
+        GroupLogEvent.objects.create(group=group, type=GroupLogEventType.STARTED)
+
+        CoordinatorLogEvent.objects.bulk_create(
+            CoordinatorLogEvent(
+                coordinator=coordinator,
+                group=group,
+                type=CoordinatorLogEventType.TOOK_NEW_GROUP,
+            )
+            for coordinator in group.coordinators.iterator()
+        )
+
+        StudentLogEvent.objects.bulk_create(
+            StudentLogEvent(student=student, to_group=group, type=StudentLogEventType.STUDY_START)
+            for student in group.students.iterator()
+        )
+
+        TeacherLogEvent.objects.bulk_create(
+            TeacherLogEvent(teacher=teacher, to_group=group, type=TeacherLogEventType.STUDY_START)
+            for teacher in group.teachers.iterator()
+        )
+
+    @classmethod
+    def _set_statuses_start(cls, group: Group) -> None:
+        timestamp = datetime.now(tz=pytz.UTC)
+
+        cls._set_status(obj=group, status=GroupStatus.WORKING, status_since=timestamp)
+
+        # Resulting status of coordinator depends on how many groups they now have.
+        # We can't use "has_enough_groups" or "has_reached_group_limit" because they are properties
+        group.coordinators.filter(groups__count_lt=CoordinatorGroupLimit.MIN).update(
```

Comment:
Will this filters used anywhere else? If so, I advise to put them to model Manager / custom queryset class, to have something like

`group.coordinators.filter_have_enough_groups().update(...)`

https://docs.djangoproject.com/en/4.2/topics/db/managers/

Reply:
Also, maybe you can do a method in custom queryset kind of `set_status(status: Status)`, which will do
`self.update(status=status, status_since=timezone.now())`
to avoid duplication of same params in these updates